libs/parallelism/executors/tests/unit/scheduler_executor.cpp

```diff
@@ -0,0 +1,300 @@
+{
+    executed = false;
+
+    hpx::parallel::execution::async_execute(exec, &test, 42).get();
+
+    HPX_TEST(executed);
+}
+
+///////////////////////////////////////////////////////////////////////////////
+void test_f(hpx::future<void> f, int passed_through)
+{
+    HPX_TEST(f.is_ready());    // make sure, future is ready
+
+    f.get();    // propagate exceptions
+
+    HPX_TEST_EQ(passed_through, 42);
+
+    executed = true;
+}
+
+template <typename Executor>
+void test_then(Executor&& exec)
+{
+    hpx::future<void> f = hpx::make_ready_future();
+
+    executed = false;
+
+    hpx::parallel::execution::then_execute(exec, &test_f, std::move(f), 42)
+        .get();
+
+    HPX_TEST(executed);
+}
+
+///////////////////////////////////////////////////////////////////////////////
+void bulk_test(int seq, int passed_through)    //-V813
+{
+    HPX_TEST_EQ(passed_through, 42);
+
+    if (seq == 0)
+    {
+        executed = true;
+    }
+}
+
+template <typename Executor>
+void test_bulk_sync(Executor&& exec)
+{
+    using hpx::util::placeholders::_1;
+    using hpx::util::placeholders::_2;
+
+    executed = false;
+
+    hpx::parallel::execution::bulk_sync_execute(
+        exec, hpx::util::bind(&bulk_test, _1, _2), 107, 42);
+
+    HPX_TEST(executed);
+
+    executed = false;
+
+    hpx::parallel::execution::bulk_sync_execute(exec, &bulk_test, 107, 42);
+
+    HPX_TEST(executed);
+}
+
+template <typename Executor>
+void test_bulk_async(Executor&& exec)
+{
+    using hpx::util::placeholders::_1;
+    using hpx::util::placeholders::_2;
+
+    executed = false;
+
+    auto result = hpx::parallel::execution::bulk_async_execute(
+        exec, hpx::util::bind(&bulk_test, _1, _2), 107, 42);
+    hpx::when_all(std::move(result)).get();
+
+    HPX_TEST(executed);
+
+    executed = false;
+
+    hpx::when_all(
+        hpx::parallel::execution::bulk_async_execute(exec, &bulk_test, 107, 42))
+        .get();
+
+    HPX_TEST(executed);
+}
+
+///////////////////////////////////////////////////////////////////////////////
+int bulk_test_f(int seq, hpx::shared_future<void> f,
+    int passed_through)    //-V813
+{
+    HPX_TEST(f.is_ready());    // make sure, future is ready
+
+    f.get();    // propagate exceptions
+
+    HPX_TEST_EQ(passed_through, 42);
+
+    if (seq == 0)
+    {
+        executed = true;
+    }
+
+    return seq;
+}
+
+template <typename Executor>
+void test_bulk_then(Executor&& exec)
+{
+    using hpx::util::placeholders::_1;
+    using hpx::util::placeholders::_2;
+    using hpx::util::placeholders::_3;
+
+    hpx::shared_future<void> f = hpx::make_ready_future();
+
+    {
+        executed = false;
+
+        auto result = hpx::parallel::execution::bulk_then_execute(
+            exec, hpx::util::bind(&bulk_test_f, _1, _2, _3), 107, f, 42)
+                          .get();
+
+        HPX_TEST(executed);
+        HPX_TEST(result.size() == 107);
+
+        int expected = 0;
+        for (auto i: result)
```

Comment:
And as a side-note, this needs formatting:
```suggestion
        for (auto i : result)
```